Added script file feature

peterdeme opened the pull request:
Resolves: [python-poetry#241](https://github.com/python-poetry/poetry/issues/241)

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.

# Feature: ability to add script files to distribution packages

This one is a long overdue, see: https://github.com/sdispater/poetry/issues/241 (June 2018)

The implementation is based on the official specification by @abn https://github.com/python-poetry/poetry/issues/2310

This is a very basic feature of setup.py See here: [docs](https://python-packaging.readthedocs.io/en/latest/command-line-scripts.html#the-scripts-keyword-argument).

Note: this is **not the same as** `entry_point`! This setuptools feature basically just copies over the given files to the virtualenv's `bin` folder so that it'll be available on the `$PATH` and you can simply invoke it.

### Example

```toml
[tools.poetry]
package = "my_package"

[tools.poetry.scripts]
migration = { source = "bin/run_db_migrations.sh", type = "file" }
setup_script = { source = "bin/run_setup_script.sh" } # type can be omitted!
```

Then:

```bash
poetry install
poetry run sh run_db_migrations.sh
poetry run sh run_setup_script.sh
```

### Testing

- I added a couple of automated tests
- I tested this **manually** with one of our repositories and it worked

peterdeme commented:
The build dies with an interesting error:

```
SolverProblemError

The current project's Python requirement (~2.7 || ^3.5) is not compatible with some of the required packages Python requirement:
  - futures requires Python >=2.6, <3, so it will not be satisfied for Python >=3.5,<4.0

Because no versions of isort match >4.3.21,<5.0.0
 and isort (4.3.21) depends on futures (*), isort (>=4.3.21,<5.0.0) requires futures (*).
Because futures (3.3.0) requires Python >=2.6, <3
 and no versions of futures match !=3.3.0, futures is forbidden.
Thus, isort is forbidden.
So, because poetry-core depends on isort (^4.3.21), version solving failed.
```

Yajo commented:
So, will these scripts be be available when installing the app? Or only in dev mode?

peterdeme commented:
@Yajo when installing the app. See “scripts” in setup.py

muppetjones commented:
Trying to push for poetry at my company, and this functionality is critical. Glad it's finally getting in!

peterdeme commented:
@muppetjones
Well, if you activate the virtualenv (or dont use virtualenv at all), you don’t need poetry run prefix. It works exactly the same way as setup.py.

muppetjones commented:
Perfect. That's exactly what I needed. Thanks!
Once this is merged, is there an estimate to when it'll be available in a release?

stephsamson commented on poetry/core/masonry/builders/builder.py:
```diff
@@ -273,6 +276,24 @@ def convert_entry_points(self):  # type: () -> dict
         return dict(result)
 
+    def convert_script_files(
+        self, base_path=None
```
It looks like `base_path` as a parameter is not used anywhere else. For what use case would it be useful that this parameter exists?

It is probably fine to drop this for now so that the path resolution logic is more straightforward and only add it in once there is a clear use case.

peterdeme replied:
Unfortunately the bdist and the sdist builder has a difference as far as base path goes. That was the only way to solve it.

peterdeme replied:
My bad, this parameter was originally used by the bdist builder.

stephsamson commented on poetry/core/masonry/builders/builder.py:
```diff
@@ -273,6 +276,24 @@ def convert_entry_points(self):  # type: () -> dict
+
+        for _, ep in self._poetry.local_config.get("scripts", {}).items():
+            if isinstance(ep, dict) and "source" in ep:
+                abs_path = Path.joinpath(base_path or self._path, ep["source"])
```
```suggestion
                abs_path = (base_path or self._path).resolve()
```

If we want to get the absolute path of an object, leveraging the `pathlib.resolve()` would be more suitable. We are also using this in the [`BuildIncludeFile`](https://github.com/python-poetry/poetry-core/blob/master/poetry/core/masonry/builders/builder.py#L299).

peterdeme replied:
Awesome, I didn’t know about that!

stephsamson replied:
Ah you're right -- I gave the wrong suggestion. It should be:

```
abs_path = (base_path or self._path / ep["source"]).resolve()
```

peterdeme replied:
Updated the code, it's pretty clean now as base path is not needed

stephsamson commented on poetry/core/masonry/builders/wheel.py:
```diff
@@ -167,6 +178,11 @@ def _write_record(self, wheel):
 
+    @property
+    def wheel_data_folder(self):  # type: () -> str
+        wheel_data_path = "{}-{}.data".format(self._package.name, self._meta.version)
+        return wheel_data_path
```
```suggestion
        return "{}-{}.data".format(self._package.name, self._meta.version)
```

peterdeme replied:
Fixed.

peterdeme commented on poetry/core/json/schemas/poetry-schema.json:
```diff
@@ -212,13 +210,23 @@
         "package-format": {
             "type": "string",
-            "enum": ["sdist", "wheel"],
+            "enum": [
```
Steph, this file was auto formatted by VS Code. I think, it’s the best to follow the rules whatever autoformatters dictate to us.

miigotu replied:
@peterdeme assuming autoformatters are standardized is where you are mistaken. Most python developers are probably not using vscode, which formats very differently than for example eclipse or pycharm.

peterdeme replied:
Ok. Reverted it.

peterdeme commented:
I managed to fix the failing build with a `poetry lock `command. It threw some error (max recursion) during the installation part 🤷‍♂️

abn commented:
@peterdeme thanks for pinging, I haven't looked at this one yet but will do so right after we finish up the 1.1.0 release. Apologies for not getting to this earlier.

peterdeme commented:
I had merge conflicts, so I rebase my branch. I also squashed my two commits into one.

roxchkplusony commented:
@abn did you mean poetry-core 1.1.0 or poetry 1.1.0? Since poetry 1.1.0 went out on Oct 1 and is on 1.1.4 now, would it be a good time to look into this again?

abn commented:
@roxchkplusony I can appreciate that this is an important feature for you. But please do understand that this project is maintained by a small group of volunteers. This means that we do not always get the amount of time we would like to invest into the project. And other priorities take precedence, for example bug fix releases.

Please rest assured that this is still on my rather long list of things to get to. And we appreciate your patience ;)

roxchkplusony commented:
Of course @abn, don't mean to pressure you. Just wanted clarity on what commitment/estimate you were making, and if it still held. Besides, what does it take to join that inner circle of a "small group of volunteers", anyhow?

abn commented:
> Just wanted clarity on what commitment/estimate you were making, and if it still held.

I am not sure keeping me "accountable" for time frame is appropriate in this context, obviously I do not work for you. :) Maybe that was not your intent, maybe you simply wanted to know if it is still something on my radar - and as I said before it very much is.

> Besides, what does it take to join that inner circle of a "small group of volunteers", anyhow?

I am unsure what you are implying by "inner circle" here, the maintainers do not encourage such a notion. Discussions typically happen on the discord server or on open issues.

You volunteer as with any open source project, start by contributing. If you want to be a maintainer, all I can suggest is to keep contributing long term, interact with the community respectfully and help triage issues. Speaking from my own experience, I contributed for over a year before I was invited to be a maintainer. That said, if you are curious about this, lets discuss this on the discord server and not pollute the discussion relevant to this change.

peterdeme commented:
Fun fact: I opened [my first PR](https://github.com/python-poetry/poetry/pull/1504) for this task more then a year ago. 🙂 
I'm losing hope.

justinmayer commented:
@peterdeme: As an open-source contributor, I have also felt discouragement when PR review takes a long time. As a maintainer, on the other hand, I often don't have enough free time to review PRs as quickly as I would like. So I have seen this from both perspectives and understand them well. Rest assured that folks are doing the best they can and will review your submission as soon as they free up sufficient time.

In the interim, `.github/workflows/tests.yml` doesn't exist any longer, so I think you'll need to rebase your changes.

Also, I see you have checked the `[✓] Updated documentation for changed code` checkbox, but I don't see any documentation about the new feature in this PR. That would seem important, since otherwise how will users know that this feature exists?

peterdeme commented:
@justinmayer thanks for the feedback. Sure, I understand that people do this in their free time and I respect it. One years still feels too long.

- tests.yml - sure, I'll do that in a minute
- docs - my previous PR to the main repo included docs. But as far as I know, **poetry-core** doesn't have documentation, does it?

justinmayer commented:
Ah, yes… Good point. I forgot that this repository has been separated from the one containing the documentation. I imagine the PR templates should be updated to account for that.

justinmayer commented on .github/workflows/tests.yml:
```diff
@@ -66,4 +66,4 @@ jobs:
       - name: Run pytest
         shell: bash
-        run: poetry run pytest -q tests
+        run: poetry run pytest -q tests
```
There seem to be some whitespace changes in this latest iteration. Namely, missing EOF newlines. Perhaps something VS Code is doing? I suggest reviewing the PR file list in GitHub to ensure there isn't any unintended whitespace changes: https://github.com/python-poetry/poetry-core/pull/40/files

peterdeme replied:
Yeah, I just realized it, goddamn VS Code man 🙂  I specifically used "Save Without Formatting" option, and it still removed the `\n`. I'll try to fix it.

justinmayer replied:
Hehe. Helpful editors being helpful. 😉

There are a few other missing newlines in the test fixtures. Probably worth addressing unless they need to be absent for some reason.

peterdeme replied:
Added a new line to the end of test_builders.py

peterdeme replied:
Nvm, if I add one, the linter fails. I guess I'll keep it as is.

justinmayer reviewed (commented):
"No newline at end of file": Not sure if these were intended or not, but might want to have a look to make sure.

justinmayer reviewed (approved):
Many thanks for all your work on this useful enhancement, Peter. 👍

peterdeme commented:
@justinmayer messed around with the line endings a little bit. The linter is all green now. Mind re-approving it?
Thank you 🙏

abn reviewed (changes requested):
@peterdeme appreciate effort you have put into this one. And apologies for the **long** delays on this one.

I have left my initial review comments.

linknum23 commented:
What happened here? How much work is left on this PR?

peterdeme commented:
@linknum23 I need to fix a few things, I didn't have time so far

mshafer1 commented:
> @linknum23 I need to fix a few things, I didn't have time so far

@peterdeme, would you be open to having some of these contributed via PR to your dev branch?

peterdeme commented:
@mshafer1 absolutely, go ahead

abn commented:
@peterdeme let me know when this is ready again for review.

lsamayoa commented:
@peterdeme is there anything I could do to help you with this ?

miigotu commented:
> @peterdeme is there anything I could do to help you with this ? 

I rebased and applied suggested changes to the PR and made another PR since it required a force push, but nobody has responded yet.

peterdeme commented on poetry/core/json/schemas/poetry-schema.json:
```diff
@@ -212,13 +210,23 @@
         "package-format": {
             "type": "string",
-            "enum": ["sdist", "wheel"],
+            "enum": [
```
Still don't understand why, but okay

peterdeme commented on poetry/core/masonry/builders/builder.py:
```diff
@@ -287,6 +290,22 @@ def convert_entry_points(self):  # type: () -> dict
+
+        for _, ep in self._poetry.local_config.get("scripts", {}).items():
+            if isinstance(ep, dict) and "source" in ep:
+                abs_path = Path.joinpath(self._path, ep["source"])
```
Added resolve. It doesn't throw an error if the file doesn't exist though

peterdeme replied:
update: resolve doesn't work with OSX and Windows.

Windows: `ValueError: 'C:\\Users\\runneradmin\\AppData\\Local\\Temp\\tmpg1ph7ks_\\complete\\bin\\script1.sh' does not start with 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmpg1ph7ks_\\complete'`

peterdeme commented on poetry/core/masonry/builders/builder.py:
```diff
@@ -287,6 +290,22 @@ def convert_entry_points(self):  # type: () -> dict
+
+        for _, ep in self._poetry.local_config.get("scripts", {}).items():
+            if isinstance(ep, dict) and "source" in ep:
+                abs_path = Path.joinpath(self._path, ep["source"])
```
Good point. Added check.

peterdeme commented on poetry/core/masonry/builders/sdist.py:
```diff
@@ -311,6 +317,10 @@ def find_files_to_add(
         }
 
+        # add script files
+        for abs_path in self.convert_script_files():
+            additional_files.add(abs_path)
```
Done

peterdeme commented on poetry/core/masonry/builders/wheel.py:
```diff
@@ -145,6 +146,16 @@ def _build(self, wheel):
+            self._add_file(
+                wheel,
+                abs_path,
+                "{}/{}/{}".format(self.wheel_data_folder, "scripts", abs_path.name),
```
Done

peterdeme commented:
@abn @miigotu @lsamayoa
I rebased and updated the PR with the suggestions. One question to Arun though.

What you're proposing to me, to change `callable` is a breaking change. I can introduce it, but that's important to note.
Secondly, this is what you mean?

```toml
extra-script = {reference = "my_package.extra:main", extras = ["time"], type = "callable"}
sh-script = {reference = "bin/script1.sh", type = "file"}
```

peterdeme commented on poetry/core/json/schemas/poetry-schema.json:
```diff
@@ -513,6 +521,9 @@
                         {
                             "$ref": "#/definitions/extra-script"
+                        },
+                        {
```
I have refactored the schema as per your requirements, split it into two: script-legacy and script-table

peterdeme commented on poetry/core/masonry/builders/builder.py:
```diff
@@ -271,6 +272,8 @@ def convert_entry_points(self):  # type: () -> dict
         for name, ep in self._poetry.local_config.get("scripts", {}).items():
             extras = ""
             if isinstance(ep, dict):
+                if "source" in ep:
```
See my comment above about making that decision

justinmayer reviewed (approved):
Nice work, Peter!

peterdeme commented:
@justinmayer The code quality linter failed, just fixed the issue (missing empty line). Pushed a fix for that.

abn commented:
Thanks for updating it @peterdeme!

> What you're proposing to me, to change `callable` is a breaking change. I can introduce it, but that's important to note.
> Secondly, this is what you mean?
> 
> ```toml
> extra-script = {reference = "my_package.extra:main", extras = ["time"], type = "callable"}
> sh-script = {reference = "bin/script1.sh", type = "file"}
> ```

Not precisely, I am thinking something like this.

```toml
extra-script = {reference = "my_package.extra:main", extras = ["time"], type = "console"}
sh-script = {reference = "bin/script1.sh", type = "file"}
```

Note the change of `callable` to `console` (alternatively `gui`). Additionally, note that in python-poetry/poetry#2310 we indicate that we wont support `extras` in the first pass. This is because there are a few details to iron out on that front, additionally it seems it is discouraged in newer tools too.

peterdeme commented:
@abn 

Refactored the PR per your suggestions.

abn reviewed (commented):
:partying_face: LGTM

miigotu commented:
Thank you! This is a big win, I'll be working on more plugins now that poetry works well for packaging my project.

hcoona commented:
May I know how to use this feature now? I'm using poetry v1.1.7

I tried to add the following section but failed to run `poetry install`.

```toml
[tool.poetry.scripts]
gen_clang_syntax_check = {source = "cmd/gen_clang_syntax_check.py", type = "file"}
```

The error message is shown as following:

```text
Installing dependencies from lock file

No dependencies to install or update

Installing the current project: bmq_dev_support (0.1.0)
  NonExistentKey

  'Key "extras" does not exist.'

  at ~/.local/lib/python3.9/site-packages/tomlkit/container.py:553 in __getitem__
      549│             key = Key(key)
      550│ 
      551│         idx = self._map.get(key, None)
      552│         if idx is None:
    → 553│             raise NonExistentKey(key)
      554│ 
      555│         if isinstance(idx, tuple):
      556│             # The item we are getting is an out of order table
      557│             # so we need a proxy to retrieve the proper objects
```

vchrombie commented:
> May I know how to use this feature now? I found no document about it.

I think the PR is merged but the feature is not yet released in any of the releases. Maybe the maintainers are planning to release this in the `1.2.x` versions, not sure tho.

hcoona commented:
Got it, it's poetry-core v1.1.0a5 not poetry v1.1.0, so poetry v1.1.7 didn't contain this feature.

My script file is running with `if __name__ == '__main__'` pattern & containing args flags. It seems that the script configuration only support running specific method?

mshafer-NI commented:
@hcoona, yes the current script tag only supports a specific method (that takes no args).

What I usually do is:
```python
def main(argv=None):
    if argv=None:
        argv = sys.argv[1:]
   ... # main program

...

if __name__ == "__main__":
    main()
```

with:
```toml
[tool.poetry.scripts]
script-foo = "foo.__main__:main"
```

this allows for:
"python -m foo", "poetry run script-foo" to both access "main"
and allows for my module (foo in example) to be imported and called via "foo.main([list of args])"

Hope this is helpful.

linknum23 commented:
Looking at the test cases and schema (in poetry_schema.json) introduced by this PR

I would expect your example to look like this:

```toml
[tool.poetry.scripts]
gen_clang_syntax_check = {reference= "cmd/gen_clang_syntax_check.py", type = "file"}
```

If gen_clang_syntax_check.py is an extra standalone script (as it appears)

OR

```toml
[tool.poetry.scripts]
gen_clang_syntax_check = {reference= "cmd.gen_clang_syntax_check", type = "console"}
```

If gen_clang_syntax_check.py is a python module

I have not had a chance to use this feature yet tho, so I could have missed something.

muxator commented:
> @hcoona, yes the current script tag only supports a specific method (that takes no args).
> 
> What I usually do is:
> 
> ```python
> def main(argv=None):
>     if argv=None:
>         argv = sys.argv[1:]
>    ... # main program
> 
> ...
> 
> if __name__ == "__main__":
>     main()
> ```
> 
> with:
> 
> ```toml
> [tool.poetry.scripts]
> script-foo = "foo.__main__:main"
> ```

@mshafer-NI (or anyone who knows), do you happen to know if there is a way of supporting exit codes from `main()`?

Basically, given a slight variation on your example program, where `main()` returns an `int` instead of `None`:
```
def main(argv=None) -> int:
    if argv=None:
        argv = sys.argv[1:]
    return 1
```

Is there a way of having the installed script relaying `main()`'s return value as exit code?
I have seen this pattern in many installed python packages, I'd want to use it too, but I fail too see how to do it with poetry.

```
if __name__ == '__main__':
    sys.exit(yourmodule.main())
```

mshafer-NI commented:
@muxator - I've solved this two different ways
1) `script-foo = "foo.__main__:main"`, `main` and `main` calls `sys.exit` on error (`__main__:main` isn't part of my api, so exiting here is fine)
2) using `click` and `click.abort` ([docs](https://click.palletsprojects.com/en/8.1.x/api/?highlight=abort#click.Abort)) on error

raxod502 commented:
Did this feature get removed?

```
[tool.poetry.scripts]
somebinary = {source = "somebinary", type = "file"}
```

yields the error:

```
The Poetry configuration is invalid:
  - [scripts.somebinary] {'source': 'somebinary', 'type': 'file'} is not valid under any of the given schemas
```

There is still documentation at https://python-poetry.org/docs/pyproject#scripts that says this key exists but the examples don't cover the usage here.

keriksson-rosenqvist commented:
Does this actually work? I can't seem to get it to run. 

I've set this in my pyproject.toml
```
[tool.poetry.scripts]
my-script = {reference = "path/to/my-script.sh", type = "file"}
```
`poetry install` runs fine but when I try to use the script I get the following:

```
$ poetry run my-script
Warning: 'my-script' is an entry point defined in pyproject.toml, but it's not installed as a script. You may get improper `sys.argv[0]`.

The support to run uninstalled scripts will be removed in a future release.

Run `poetry install` to resolve and get rid of this message.


'callable'
```

What am I doing wrong and could the [Poetry documentation](https://python-poetry.org/docs/pyproject/#scripts) be updated with some examples for different script types?

wohali commented:
This PR doesn't give us what we want. See https://github.com/python-poetry/poetry/issues/2310#issuecomment-1803757408 for a detailed explanation.
